data/transactions/logic/assembler.go

```diff
@@ -1698,39 +1704,59 @@ func (ops *OpStream) cycle(macro string, previous ...string) bool {
 	return false
 }
 
-// Once a version is obtained, versionedMacroFilter goes through previously defined macros
+// Once a version is obtained, versionedCheckMacroNames goes through previously defined macros
 // and ensures they don't use opcodes/fields from the obtained version
-func (ops *OpStream) versionedMacroFilter() error {
-	var err error = nil
+func (ops *OpStream) versionedCheckMacroNames() error {
+	var errored bool
 	for macroName := range ops.macros {
 		if fieldNames[ops.Version][macroName] {
-			err = fmt.Errorf("%s is defined as a macro but is a field name in version %d", macroName, ops.Version)
+			ops.errorf("%s is defined as a macro but is a field name in version %d", macroName, ops.Version)
+			errored = true
+			delete(ops.macros, macroName)
 		} else if _, ok := OpsByName[ops.Version][macroName]; ok {
-			err = fmt.Errorf("%s is defined as a macro but is an opcode in version %d", macroName, ops.Version)
-		} else {
-			continue
+			ops.errorf("%s is defined as a macro but is an opcode in version %d", macroName, ops.Version)
+			errored = true
+			delete(ops.macros, macroName)
 		}
-		ops.error(err)
-		delete(ops.macros, macroName)
 	}
-	if err != nil {
-		err = errors.New("version is incompatible with defined macros")
+	if errored {
+		return errors.New("version is incompatible with defined macros")
 	}
-	return err
+	return nil
 }
 
-func (ops *OpStream) filterMacro(macroName string) error {
+func checkMacroName(macroName string, version uint64) error {
+	otherAllowedChars := [256]bool{'+': true, '-': true, '*': true, '/': true, '^': true, '%': true, '&': true, '|': true, '#': true, '!': true, '>': true, '<': true, ':': true, '=': true, '?': true}
+	for i := 0; i < len(macroName); i++ {
+		c := macroName[i]
+		if !unicode.IsLetter(rune(c)) && !unicode.IsDigit(rune(c)) && !otherAllowedChars[c] {
+			return fmt.Errorf("%s character not allowed in macro name", string(c))
+		}
+	}
+	firstChar := macroName[0]
+	lastChar := macroName[len(macroName)-1]
+	if firstChar == '-' && len(macroName) > 1 {
+		if unicode.IsDigit(rune(macroName[1])) {
+			return fmt.Errorf("Cannot begin macro name with number: %s", macroName)
+		}
+	}
+	if lastChar == ':' {
+		return fmt.Errorf("Macro names cannot end in colons: %s", macroName)
+	}
+	if macroName == "b64" || macroName == "base64" {
```

Comment:
Filter `b32` and `base32` as well?

Reply:
We don't have b32 syntax in teal, do we?

Reply:
We do! https://github.com/algorand/go-algorand/blob/feab0180dc0f7bc1ed75e85b858fa5b9975c3414/data/transactions/logic/assembler.go#L563

Reply:
Oh. Weird. The "lexer" doesn't know about that. So I think b32 must only work as b32(...) while b64 works as b64 (...) Which I think means we don't need to exclude b32 as a macro. b32(...) is one token.

Reply:
Looking at parseBinaryArgs, it appears we do in fact allow you to do b32 ... and base32 ... with spaces instead of (). Good find Jason!

The lexer I guess does not care about b32 b/c you wouldn't have a / in b32 and it only tracks b64 to avoid // being treated as a comment inside

Reply:
Weird. To me it looks like b32 is definitely meant to be used in the same way, so that seems like an oversight https://github.com/algorand/go-algorand/blob/feab0180dc0f7bc1ed75e85b858fa5b9975c3414/data/transactions/logic/assembler.go#L593

Reply:
Ah, that's why it cares. Makes sense, thanks.